src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,111 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIALIZE_ELEMENT_DATA_NUMBER = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[INITIALIZE_ELEMENT_DATA_NUMBER];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        growIfArrayFull();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        validateIndexToAdd(index);
+        growIfArrayFull();
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndexToRemove(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        validateIndexToRemove(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        validateIndexToRemove(index);
+        T elementToRemove = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        size--;
+        return elementToRemove;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = findIndexByValue(element);
+        validateIndexToRemove(index);
+        System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        size--;
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private int findIndexByValue(T value) {
+        int i;
+        for (i = 0; i < size; i++) {
+            if ((value == null && elementData[i] == null)
+                    || (value != null && value.equals(elementData[i]))) {
```

Comment:
```suggestion
            if (value == null && elementData[i] == null
                    || value != null && value.equals(elementData[i])) {
```
these brackets are redundant https://introcs.cs.princeton.edu/java/11precedence/